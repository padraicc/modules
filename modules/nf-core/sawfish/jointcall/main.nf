process SAWFISH_JOINTCALL {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/f8/f8f28fcf0823d39a8038b02cc5ff86d7cfbcf92e6f5c8540c25e110c6eb1f9be/data' :
        'community.wave.seqera.io/library/sawfish:2.1.1--1fd57d06a7186245' }"

    input:
    tuple val(meta), path(sample_dirs)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(bams), path(bais)
    tuple val(meta4), path(sample_csv)

    output:
    tuple val(meta), path("joint_call_dir/*_genotyped.sv.vcf.gz")     , emit: vcf
    tuple val(meta), path("joint_call_dir/*_genotyped.sv.vcf.gz.tbi") , emit: tbi
    tuple val(meta), path("joint_call_dir/contig.alignment.bam")      , emit: bam
    tuple val(meta), path("joint_call_dir/contig.alignment.bam.csi")  , emit: bam_index
    tuple val(meta), path("joint_call_dir/run.stats.json")            , emit: stats
    tuple val(meta), path("joint_call_dir/samples/")                  , emit: samples_dir
    tuple val(meta), path("joint_call_dir/sawfish.log")               , emit: log
    path "versions.yml"                                               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def sample_args = sample_csv ? '' : sample_dirs.collect { "--sample ${it}" }.join(' ')
    def sample_csv_arg = sample_csv ? "--sample-csv ${sample_csv}" : ""

    """
    sawfish \\
        joint-call \\
        --threads $task.cpus \\
        --ref $fasta \\
        $sample_args \\
        $args \\
        $sample_csv_arg \\
        --output-dir joint_call_dir

    # Rename the output files to include prefix
    if [ -f joint_call_dir/genotyped.sv.vcf.gz ]; then
        mv joint_call_dir/genotyped.sv.vcf.gz joint_call_dir/${prefix}_genotyped.sv.vcf.gz
    fi
    if [ -f joint_call_dir/genotyped.sv.vcf.gz.tbi ]; then
        mv joint_call_dir/genotyped.sv.vcf.gz.tbi joint_call_dir/${prefix}_genotyped.sv.vcf.gz.tbi
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sawfish: \$(sawfish --version | sed 's/sawfish //g')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p joint_call_dir
    mkdir -p joint_call_dir/samples/sample0001_test/
    mkdir -p joint_call_dir/samples/sample0002_test/
    echo "##fileformat=VCFv4.4" | gzip > joint_call_dir/${prefix}_genotyped.sv.vcf.gz
    echo "VCF index content" > joint_call_dir/${prefix}_genotyped.sv.vcf.gz.tbi
    echo "BAM content" > joint_call_dir/contig.alignment.bam
    echo "BAM index content" > joint_call_dir/contig.alignment.bam.csi
    echo '{"run_stats": "mock"}' > joint_call_dir/run.stats.json
    echo "Log content" > joint_call_dir/sawfish.log
    echo "Copy number bedgraph content" > joint_call_dir/samples/sample0001_test/copynum.bedgraph
    echo "Depth content" > joint_call_dir/samples/sample0001_test/depth.bw
    echo "GC bias corrected depth content" > joint_call_dir/samples/sample0001_test/gc_bias_corrected_depth.bw
    echo "CN summary content" > joint_call_dir/samples/sample0001_test/copynum.summary.json
    echo "Copy number bedgraph content" > joint_call_dir/samples/sample0002_test/copynum.bedgraph
    echo "Depth content" > joint_call_dir/samples/sample0002_test/depth.bw
    echo "GC bias corrected depth content" > joint_call_dir/samples/sample0002_test/gc_bias_corrected_depth.bw
    echo "CN summary content" > joint_call_dir/samples/sample0002_test/copynum.summary.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sawfish: \$(sawfish --version | sed 's/sawfish //g')
    END_VERSIONS
    """
}
