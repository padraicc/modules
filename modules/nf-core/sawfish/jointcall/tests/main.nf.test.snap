{
    "sawfish_jointcall": {
        "content": [
            [
                "versions.yml:md5,a4558bd0e8487faf7562a43cc1b8de9c"
            ],
            "test_genotyped.sv.vcf.gz.tbi",
            [
                "##reference=file://genome3.fasta",
                "##source=\"sawfish 2.1.1\"",
                "##sawfish_cmdline=\"sawfish joint-call --threads 2 --ref genome3.fasta --sample test_discover_dir --output-dir joint_call_dir\"",
                "##contig=<ID=chr19:45760000-45770300,length=10301>",
                "##ALT=<ID=DEL,Description=\"Deletion\">",
                "##ALT=<ID=INS,Description=\"Insertion\">",
                "##ALT=<ID=DUP,Description=\"Duplication\">",
                "##ALT=<ID=DUP:TANDEM,Description=\"Tandem Duplication\">",
                "##ALT=<ID=INV,Description=\"Inversion\">",
                "##ALT=<ID=CNV,Description=\"Copy number variable region\">",
                "##FILTER=<ID=PASS,Description=\"All filters passed\">",
                "##FILTER=<ID=.,Description=\"Unknown filtration status\">",
                "##FILTER=<ID=InvBreakpoint,Description=\"Breakpoint represented as part of an inversion record (same EVENT ID)\">",
                "##FILTER=<ID=MinQUAL,Description=\"QUAL score is less than 10\">",
                "##FILTER=<ID=MaxScoringDepth,Description=\"SV candidate exceeds max scoring depth in at least one sample\">",
                "##FILTER=<ID=ConflictingBreakpointGT,Description=\"Genotypes of breakpoints in a multi-breakpoint event conflict in the majority of cases\">",
                "##INFO=<ID=END,Number=1,Type=Integer,Description=\"End position of the variant described in this record\">",
                "##INFO=<ID=EVENT,Number=A,Type=String,Description=\"ID of associated event\">",
                "##INFO=<ID=EVENTTYPE,Number=A,Type=String,Description=\"Type of associated event\">",
                "##INFO=<ID=HOMLEN,Number=.,Type=Integer,Description=\"Length of base pair identical micro-homology at event breakpoints\">",
                "##INFO=<ID=HOMSEQ,Number=.,Type=String,Description=\"Sequence of base pair identical micro-homology at event breakpoints\">",
                "##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description=\"Imprecise structural variation\">",
                "##INFO=<ID=INSLEN,Number=.,Type=Integer,Description=\"Insertion length\">",
                "##INFO=<ID=INSSEQ,Number=.,Type=String,Description=\"Insertion sequence (for symbolic alt alleles)\">",
                "##INFO=<ID=MATEID,Number=.,Type=String,Description=\"ID of mate breakends\">",
                "##INFO=<ID=SVCLAIM,Number=A,Type=String,Description=\"Claim made by the structural variant call. Valid values are D, J, DJ for abundance, adjacency and both respectively\">",
                "##INFO=<ID=SVLEN,Number=A,Type=Integer,Description=\"Length of structural variant\">",
                "##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">",
                "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">",
                "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">",
                "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Phred-scaled genotype likelihoods rounded to the closest integer\">",
                "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Read depth for each allele\">",
                "##FORMAT=<ID=CN,Number=1,Type=Float,Description=\"Copy number\">",
                "##FORMAT=<ID=CNQ,Number=1,Type=Float,Description=\"Copy number genotype quality\">",
                "##FORMAT=<ID=PS,Number=1,Type=Integer,Description=\"Phase set identifier\">",
                "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNA03697B2",
                "chr19:45760000-45770300\t1\tsawfish:CNV:0:0:0\tC\t<DEL>\t62\tPASS\tIMPRECISE;SVTYPE=DEL;END=9000;SVLEN=8999;SVCLAIM=D\tGT:CN:CNQ\t1/1:0:31",
                "chr19:45760000-45770300\t10205\tsawfish:0:0:0:0\tC\tCCCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAG\t999\tPASS\tSVTYPE=INS;END=10205;SVLEN=1633;HOMLEN=1;HOMSEQ=C;INSLEN=1633\tGT:GQ:PL:AD\t1/1:129:999,129,0:0,43"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-11T10:54:28.528423404"
    },
    "sawfish_jointcall_stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_genotyped.sv.vcf.gz:md5,fe04369855e9af3a4b31b5e2c5f7b067"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_genotyped.sv.vcf.gz.tbi:md5,4fe64addde9dfff20332c4bd8a10c008"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "contig.alignment.bam:md5,181e94e83c5a12dfe14d76208f1dae3d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "contig.alignment.bam.csi:md5,8ba1cf883b6c208fa9cfe6631269a464"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "run.stats.json:md5,ad3904d833cd372275bc4f4dc5707ea3"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            [
                                "copynum.bedgraph:md5,df4f470e658adbbd0ae2942cf88eab2c",
                                "copynum.summary.json:md5,1a56934708910974573365e8367cdd46",
                                "depth.bw:md5,6daa4b2126b0638a8f5274b391b8b8f7",
                                "gc_bias_corrected_depth.bw:md5,0d02346d621433de25138fa726cfd273"
                            ],
                            [
                                "copynum.bedgraph:md5,df4f470e658adbbd0ae2942cf88eab2c",
                                "copynum.summary.json:md5,1a56934708910974573365e8367cdd46",
                                "depth.bw:md5,6daa4b2126b0638a8f5274b391b8b8f7",
                                "gc_bias_corrected_depth.bw:md5,0d02346d621433de25138fa726cfd273"
                            ]
                        ]
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "sawfish.log:md5,50b046cd550508b1a438eb088a15502a"
                    ]
                ],
                "7": [
                    "versions.yml:md5,a4558bd0e8487faf7562a43cc1b8de9c"
                ],
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "contig.alignment.bam:md5,181e94e83c5a12dfe14d76208f1dae3d"
                    ]
                ],
                "bam_index": [
                    [
                        {
                            "id": "test"
                        },
                        "contig.alignment.bam.csi:md5,8ba1cf883b6c208fa9cfe6631269a464"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test"
                        },
                        "sawfish.log:md5,50b046cd550508b1a438eb088a15502a"
                    ]
                ],
                "samples_dir": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            [
                                "copynum.bedgraph:md5,df4f470e658adbbd0ae2942cf88eab2c",
                                "copynum.summary.json:md5,1a56934708910974573365e8367cdd46",
                                "depth.bw:md5,6daa4b2126b0638a8f5274b391b8b8f7",
                                "gc_bias_corrected_depth.bw:md5,0d02346d621433de25138fa726cfd273"
                            ],
                            [
                                "copynum.bedgraph:md5,df4f470e658adbbd0ae2942cf88eab2c",
                                "copynum.summary.json:md5,1a56934708910974573365e8367cdd46",
                                "depth.bw:md5,6daa4b2126b0638a8f5274b391b8b8f7",
                                "gc_bias_corrected_depth.bw:md5,0d02346d621433de25138fa726cfd273"
                            ]
                        ]
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "run.stats.json:md5,ad3904d833cd372275bc4f4dc5707ea3"
                    ]
                ],
                "tbi": [
                    [
                        {
                            "id": "test"
                        },
                        "test_genotyped.sv.vcf.gz.tbi:md5,4fe64addde9dfff20332c4bd8a10c008"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_genotyped.sv.vcf.gz:md5,fe04369855e9af3a4b31b5e2c5f7b067"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a4558bd0e8487faf7562a43cc1b8de9c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-11T11:26:43.621089258"
    }
}